File: flytekit-java/src/main/java/org/flyte/flytekit/SdkBindingDatas.java
/*
 * Copyright 2021 Flyte Authors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
package org.flyte.flytekit;

import java.time.Duration;
import java.time.Instant;
import java.time.LocalDate;
import java.time.ZoneOffset;
import java.util.List;
import java.util.Map;
import org.flyte.api.v1.LiteralType;

public final class SdkBindingDatas {

  private SdkBindingDatas() {
    // prevent instantiation
  }

  /**
   * Creates a {@code SdkBindingData} for a flyte integer ({@link Long} for java) with the given
   * value.
   *
   * @param value the simple value for this data
   * @return the new {@code SdkBindingData}
   */
  public static SdkBindingData<Long> ofInteger(long value) {
    return SdkBindingData.literal(SdkLiteralTypes.integers(), value);
  }

  /**
   * Creates a {@code SdkBindingData} for a flyte float ({@link Double} for java) with the given
   * value.
   *
   * @param value the simple value for this data
   * @return the new {@code SdkBindingData}
   */
  public static SdkBindingData<Double> ofFloat(double value) {
    return SdkBindingData.literal(SdkLiteralTypes.floats(), value);
  }

  /**
   * Creates a {@code SdkBindingData} for a flyte String with the given value.
   *
   * @param value the simple value for this data
   * @return the new {@code SdkBindingData}
   */
  public static SdkBindingData<String> ofString(String value) {
    return SdkBindingData.literal(SdkLiteralTypes.strings(), value);
  }

  /**
   * Creates a {@code SdkBindingData} for a flyte boolean with the given value.
   *
   * @param value the simple value for this data
   * @return the new {@code SdkBindingData}
   */
  public static SdkBindingData<Boolean> ofBoolean(boolean value) {
    return SdkBindingData.literal(SdkLiteralTypes.booleans(), value);
  }

  /**
   * Creates a {@code SdkBindingData} for a flyte Datetime ({@link Instant} for java) with the given
   * date at 00:00 on UTC.
   *
   * @param year the year to represent, from {@code Year.MIN_VALUE} to {@code Year.MAX_VALUE}
   * @param month the month-of-year to represent, from 1 (January) to 12 (December)
   * @param day the day-of-month to represent, from 1 to 31
   * @return the new {@code SdkBindingData}
   */
  public static SdkBindingData<Instant> ofDatetime(int year, int month, int day) {
    Instant instant = LocalDate.of(year, month, day).atStartOfDay().toInstant(ZoneOffset.UTC);
    return ofDatetime(instant);
  }

  /**
   * Creates a {@code SdkBindingData} for a flyte Datetime ({@link Instant} for java) with the given
   * value.
   *
   * @param value the simple value for this data
   * @return the new {@code SdkBindingData}
   */
  public static SdkBindingData<Instant> ofDatetime(Instant value) {
    return SdkBindingData.literal(SdkLiteralTypes.datetimes(), value);
  }

  /**
   * Creates a {@code SdkBindingData} for a flyte Duration for java with the given value.
   *
   * @param value the simple value for this data
   * @return the new {@code SdkBindingData}
   */
  public static SdkBindingData<Duration> ofDuration(Duration value) {
    return SdkBindingData.literal(SdkLiteralTypes.durations(), value);
  }

  /**
   * Creates a {@code SdkBindingData} for a flyte collection given a java {@code List<T>} and the
   * elements type.
   *
   * @param elementType a {@link SdkLiteralType} for the collection elements type.
   * @param collection collection to represent on this data.
   * @return the new {@code SdkBindingData}
   */
  public static <T> SdkBindingData<List<T>> ofCollection(
      SdkLiteralType<T> elementType, List<T> collection) {
    return SdkBindingData.literal(SdkLiteralTypes.collections(elementType), collection);
  }

  /**
   * Creates a {@code SdkBindingData} for a flyte collection of string given a java {@code
   * List<String>}.
   *
   * @param collection collection to represent on this data.
   * @return the new {@code SdkBindingData}
   */
  public static SdkBindingData<List<String>> ofStringCollection(List<String> collection) {
    return ofCollection(SdkLiteralTypes.strings(), collection);
  }

  /**
   * Creates a {@code SdkBindingData} for a flyte collection of float given a java {@code
   * List<Double>}.
   *
   * @param collection collection to represent on this data.
   * @return the new {@code SdkBindingData}
   */
  public static SdkBindingData<List<Double>> ofFloatCollection(List<Double> collection) {
    return ofCollection(SdkLiteralTypes.floats(), collection);
  }

  /**
   * Creates a {@code SdkBindingData} for a flyte collection of integer given a java {@code
   * List<Long>}.
   *
   * @param collection collection to represent on this data.
   * @return the new {@code SdkBindingData}
   */
  public static SdkBindingData<List<Long>> ofIntegerCollection(List<Long> collection) {
    return ofCollection(SdkLiteralTypes.integers(), collection);
  }

  /**
   * Creates a {@code SdkBindingData} for a flyte collection of boolean given a java {@code
   * List<Boolean>}.
   *
   * @param collection collection to represent on this data.
   * @return the new {@code SdkBindingData}
   */
  public static SdkBindingData<List<Boolean>> ofBooleanCollection(List<Boolean> collection) {
    return ofCollection(SdkLiteralTypes.booleans(), collection);
  }

  /**
   * Creates a {@code SdkBindingData} for a flyte collection of Duration given a java {@code
   * List<Duration>}.
   *
   * @param collection collection to represent on this data.
   * @return the new {@code SdkBindingData}
   */
  public static SdkBindingData<List<Duration>> ofDurationCollection(List<Duration> collection) {
    return ofCollection(SdkLiteralTypes.durations(), collection);
  }

  /**
   * Creates a {@code SdkBindingData} for a flyte collection of datetime given a java {@code
   * List<Instant>}.
   *
   * @param collection collection to represent on this data.
   * @return the new {@code SdkBindingData}
   */
  public static SdkBindingData<List<Instant>> ofDatetimeCollection(List<Instant> collection) {
    return ofCollection(SdkLiteralTypes.datetimes(), collection);
  }

  /**
   * Creates a {@code SdkBindingData} for a flyte map given a java {@code Map<String, T>} and a
   * function to know how to convert each entry values form such map to a {@code SdkBindingData}.
   *
   * @param valuesType literal type for the values of the map, keys are always strings.
   * @param map map to represent on this data.
   * @return the new {@code SdkBindingData}
   */
  public static <T> SdkBindingData<Map<String, T>> ofMap(
      SdkLiteralType<T> valuesType, Map<String, T> map) {
    return SdkBindingData.literal(SdkLiteralTypes.maps(valuesType), map);
  }

  /**
   * Creates a {@code SdkBindingData} for a flyte map of string given a java {@code Map<String,
   * String>}.
   *
   * @param map map to represent on this data.
   * @return the new {@code SdkBindingData}
   */
  public static SdkBindingData<Map<String, String>> ofStringMap(Map<String, String> map) {
    return ofMap(SdkLiteralTypes.strings(), map);
  }

  /**
   * Creates a {@code SdkBindingData} for a flyte map of float given a java {@code Map<String,
   * Double>}.
   *
   * @param map map to represent on this data.
   * @return the new {@code SdkBindingData}
   */
  public static SdkBindingData<Map<String, Double>> ofFloatMap(Map<String, Double> map) {
    return ofMap(SdkLiteralTypes.floats(), map);
  }

  /**
   * Creates a {@code SdkBindingData} for a flyte map of integer given a java {@code Map<String,
   * Long>}.
   *
   * @param map map to represent on this data.
   * @return the new {@code SdkBindingData}
   */
  public static SdkBindingData<Map<String, Long>> ofIntegerMap(Map<String, Long> map) {
    return ofMap(SdkLiteralTypes.integers(), map);
  }

  /**
   * Creates a {@code SdkBindingData} for a flyte map of boolean given a java {@code Map<String,
   * Boolean>}.
   *
   * @param map map to represent on this data.
   * @return the new {@code SdkBindingData}
   */
  public static SdkBindingData<Map<String, Boolean>> ofBooleanMap(Map<String, Boolean> map) {
    return ofMap(SdkLiteralTypes.booleans(), map);
  }

  /**
   * Creates a {@code SdkBindingData} for a flyte map of duration given a java {@code Map<String,
   * Duration>}.
   *
   * @param map map to represent on this data.
   * @return the new {@code SdkBindingData}
   */
  public static SdkBindingData<Map<String, Duration>> ofDurationMap(Map<String, Duration> map) {
    return ofMap(SdkLiteralTypes.durations(), map);
  }

  /**
   * Creates a {@code SdkBindingData} for a flyte map of datetime given a java {@code Map<String,
   * Instant>}.
   *
   * @param map map to represent on this data.
   * @return the new {@code SdkBindingData}
   */
  public static SdkBindingData<Map<String, Instant>> ofDatetimeMap(Map<String, Instant> map) {
    return ofMap(SdkLiteralTypes.datetimes(), map);
  }

  /**
   * Creates a {@code SdkBindingData} for a flyte collection given a java {@code
   * List<SdkBindingData<T>>} and a literalType tp be used.
   *
   * @param elements collection to represent on this data.
   * @param elementType literal type for the elements in the collection.
   * @return the new {@code SdkBindingData}
   */
  public static <T> SdkBindingData<List<T>> ofBindingCollection(
      SdkLiteralType<T> elementType, List<SdkBindingData<T>> elements) {
    return SdkBindingData.bindingCollection(elementType, elements);
  }

  /**
   * Creates a {@code SdkBindingData} for a flyte map given a java {@code Map<String,
   * SdkBindingData<T>>} and a literalType tp be used.
   *
   * @param valueMap collection to represent on this data.
   * @param valuesType literal type for the whole map. It must be a {@link
   *     LiteralType.Kind#MAP_VALUE_TYPE}.
   * @return the new {@code SdkBindingData}
   */
  public static <T> SdkBindingData<Map<String, T>> ofBindingMap(
      SdkLiteralType<T> valuesType, Map<String, SdkBindingData<T>> valueMap) {

    return SdkBindingData.bindingMap(valuesType, valueMap);
  }
}

File: flytekit-scala_2.13/src/main/scala/org/flyte/flytekitscala/SdkBindingDatas.scala
/*
 * Copyright 2021 Flyte Authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
package org.flyte.flytekitscala

import org.flyte.api.v1.{
  BindingData,
  LiteralType,
  Primitive,
  Scalar,
  SimpleType
}
import org.flyte.flytekit.SdkBindingData

import java.time.{Duration, Instant}
import scala.collection.JavaConverters._

/** Utility to create [[SdkBindingData]] using scala raw types.
  */
object SdkBindingDatas {

  /** Creates a [[SdkBindingData]] for a flyte string ([[String]] for scala)
    * with the given value.
    *
    * @param string
    *   the simple value for this data
    * @return
    *   the new {[[SdkBindingData]]
    */
  def ofString(string: String): SdkBindingData[String] =
    createSdkBindingData(string)

  /** Creates a [[SdkBindingData]] for a flyte integer ([[Long]] for scala) with
    * the given value.
    *
    * @param long
    *   the simple value for this data
    * @return
    *   the new {[[SdkBindingData]]
    */
  def ofInteger(long: Long): SdkBindingData[Long] =
    createSdkBindingData(long)

  /** Creates a [[SdkBindingData]] for a flyte float ([[Double]] for scala) with
    * the given value.
    *
    * @param double
    *   the simple value for this data
    * @return
    *   the new {[[SdkBindingData]]
    */
  def ofFloat(double: Double): SdkBindingData[Double] =
    createSdkBindingData(double)

  /** Creates a [[SdkBindingData]] for a flyte boolean ([[Boolean]] for scala)
    * with the given value.
    *
    * @param boolean
    *   the simple value for this data
    * @return
    *   the new {[[SdkBindingData]]
    */
  def ofBoolean(
      boolean: Boolean
  ): SdkBindingData[Boolean] =
    createSdkBindingData(boolean)

  /** Creates a [[SdkBindingData]] for a flyte instant ([[Instant]] for scala)
    * with the given value.
    *
    * @param instant
    *   the simple value for this data
    * @return
    *   the new {[[SdkBindingData]]
    */
  def ofDateTime(instant: Instant): SdkBindingData[Instant] =
    createSdkBindingData(instant)

  /** Creates a [[SdkBindingData]] for a flyte duration ([[Duration]] for scala)
    * with the given value.
    *
    * @param duration
    *   the simple value for this data
    * @return
    *   the new {[[SdkBindingData]]
    */
  def ofDuration(
      duration: Duration
  ): SdkBindingData[Duration] = createSdkBindingData(duration)

  /** Creates a [[SdkBindingData]] for a flyte collection given a scala
    * [[List]].
    *
    * @param collection
    *   collection to represent on this data.
    * @return
    *   the new [[SdkBindingData]]
    */
  def ofCollection[T](
      collection: List[T]
  ): SdkBindingData[List[T]] = createSdkBindingData(collection)

  /** Creates a [[SdkBindingData]] for a flyte collection given a scala
    * [[List]].
    *
    * @param literalType
    *   literal type for the whole collection. It must be a
    *   [[LiteralType.Kind.COLLECTION_TYPE]].
    * @param collection
    *   collection to represent on this data.
    * @return
    *   the new [[SdkBindingData]]
    */
  def ofCollection[T](
      literalType: LiteralType,
      collection: List[T]
  ): SdkBindingData[List[T]] =
    createSdkBindingData(collection, Option(literalType))

  /** Creates a [[SdkBindingDatas]] for a flyte string collection given a scala
    * [[List]].
    *
    * @param collection
    *   collection to represent on this data.
    * @return
    *   the new [[SdkBindingData]]
    */
  def ofStringCollection(
      collection: List[String]
  ): SdkBindingData[List[String]] =
    createSdkBindingData(
      collection,
      Option(
        LiteralType.ofCollectionType(
          LiteralType.ofSimpleType(SimpleType.STRING)
        )
      )
    )

  /** Creates a [[SdkBindingData]] for a flyte integer collection given a scala
    * [[List]].
    *
    * @param collection
    *   collection to represent on this data.
    * @return
    *   the new [[SdkBindingDatas]]
    */
  def ofIntegerCollection(
      collection: List[Long]
  ): SdkBindingData[List[Long]] =
    createSdkBindingData(
      collection,
      Option(
        LiteralType.ofCollectionType(
          LiteralType.ofSimpleType(SimpleType.INTEGER)
        )
      )
    )

  /** Creates a [[SdkBindingData]] for a flyte boolean collection given a scala
    * [[List]].
    *
    * @param collection
    *   collection to represent on this data.
    * @return
    *   the new [[SdkBindingData]]
    */
  def ofBooleanCollection(
      collection: List[Boolean]
  ): SdkBindingData[List[Boolean]] =
    createSdkBindingData(
      collection,
      Option(
        LiteralType.ofCollectionType(
          LiteralType.ofSimpleType(SimpleType.BOOLEAN)
        )
      )
    )

  /** Creates a [[SdkBindingData]] for a flyte float collection given a scala
    * [[List]].
    *
    * @param collection
    *   collection to represent on this data.
    * @return
    *   the new [[SdkBindingData]]
    */
  def ofFloatCollection(
      collection: List[Double]
  ): SdkBindingData[List[Double]] =
    createSdkBindingData(
      collection,
      Option(
        LiteralType.ofCollectionType(LiteralType.ofSimpleType(SimpleType.FLOAT))
      )
    )

  /** Creates a [[SdkBindingData]] for a flyte datetime collection given a scala
    * [[List]].
    *
    * @param collection
    *   collection to represent on this data.
    * @return
    *   the new [[SdkBindingData]]
    */
  def ofInstantCollection(
      collection: List[Instant]
  ): SdkBindingData[List[Instant]] =
    createSdkBindingData(
      collection,
      Option(
        LiteralType.ofCollectionType(
          LiteralType.ofSimpleType(SimpleType.DATETIME)
        )
      )
    )

  /** Creates a [[SdkBindingData]] for a flyte duration collection given a scala
    * [[List]].
    *
    * @param collection
    *   collection to represent on this data.
    * @return
    *   the new [[SdkBindingData]]
    */
  def ofDurationCollection(
      collection: List[Duration]
  ): SdkBindingData[List[Duration]] =
    createSdkBindingData(
      collection,
      Option(
        LiteralType.ofCollectionType(
          LiteralType.ofSimpleType(SimpleType.DURATION)
        )
      )
    )

  /** Creates a [[SdkBindingData]] for a flyte map given a scala [[Map]].
    *
    * @param map
    *   collection to represent on this data.
    * @return
    *   the new [[SdkBindingData]]
    */
  def ofMap[T](
      map: Map[String, T]
  ): SdkBindingData[Map[String, T]] = createSdkBindingData(map)

  /** Creates a [[SdkBindingData]] for a flyte string map given a scala [[Map]].
    *
    * @param map
    *   collection to represent on this data.
    * @return
    *   the new [[SdkBindingDatas]]
    */
  def ofStringMap(
      map: Map[String, String]
  ): SdkBindingData[Map[String, String]] =
    createSdkBindingData(
      map,
      Option(
        LiteralType.ofMapValueType(LiteralType.ofSimpleType(SimpleType.STRING))
      )
    )

  /** Creates a [[SdkBindingData]] for a flyte long map given a scala [[Map]].
    *
    * @param map
    *   collection to represent on this data.
    * @return
    *   the new [[SdkBindingData]]
    */
  def ofIntegerMap(
      map: Map[String, Long]
  ): SdkBindingData[Map[String, Long]] =
    createSdkBindingData(
      map,
      Option(
        LiteralType.ofMapValueType(LiteralType.ofSimpleType(SimpleType.INTEGER))
      )
    )

  /** Creates a [[SdkBindingData]] for a flyte boolean map given a scala
    * [[Map]].
    *
    * @param map
    *   collection to represent on this data.
    * @return
    *   the new [[SdkBindingData]]
    */
  def ofBooleanMap(
      map: Map[String, Boolean]
  ): SdkBindingData[Map[String, Boolean]] =
    createSdkBindingData(
      map,
      Option(
        LiteralType.ofMapValueType(LiteralType.ofSimpleType(SimpleType.BOOLEAN))
      )
    )

  /** Creates a [[SdkBindingData]] for a flyte double map given a scala [[Map]].
    *
    * @param map
    *   collection to represent on this data.
    * @return
    *   the new [[SdkBindingData]]
    */
  def ofFloatMap(
      map: Map[String, Double]
  ): SdkBindingData[Map[String, Double]] =
    createSdkBindingData(
      map,
      Option(
        LiteralType.ofMapValueType(LiteralType.ofSimpleType(SimpleType.FLOAT))
      )
    )

  /** Creates a [[SdkBindingData]] for a flyte instant map given a scala
    * [[Map]].
    *
    * @param map
    *   collection to represent on this data.
    * @return
    *   the new [[SdkBindingData]]
    */
  def ofInstantMap(
      map: Map[String, Instant]
  ): SdkBindingData[Map[String, Instant]] =
    createSdkBindingData(
      map,
      Option(
        LiteralType.ofMapValueType(
          LiteralType.ofSimpleType(SimpleType.DATETIME)
        )
      )
    )

  /** Creates a [[SdkBindingData]] for a flyte duration map given a scala
    * [[Map]].
    *
    * @param map
    *   collection to represent on this data.
    * @return
    *   the new [[SdkBindingData]]
    */
  def ofDurationMap(
      map: Map[String, Duration]
  ): SdkBindingData[Map[String, Duration]] =
    createSdkBindingData(
      map,
      Option(
        LiteralType.ofMapValueType(
          LiteralType.ofSimpleType(SimpleType.DURATION)
        )
      )
    )

  /** Creates a [[SdkBindingData]] for a flyte duration map given a scala
    * [[Map]].
    *
    * @param literalType
    *   literal type for the whole collection. It must be a
    *   [[LiteralType.Kind.MAP_VALUE_TYPE]].
    * @param map
    *   collection to represent on this data.
    * @return
    *   the new [[SdkBindingData]]
    */
  def ofMap[T](
      literalType: LiteralType,
      map: Map[String, T]
  ): SdkBindingData[Map[String, T]] =
    createSdkBindingData(map, Option(literalType))

  private def toBindingData(
      value: Any,
      literalTypeOpt: Option[LiteralType]
  ): (BindingData, LiteralType) = {
    value match {
      case string: String =>
        (
          BindingData.ofScalar(
            Scalar.ofPrimitive(Primitive.ofStringValue(string))
          ),
          LiteralType.ofSimpleType(SimpleType.STRING)
        )
      case boolean: Boolean =>
        (
          BindingData.ofScalar(
            Scalar.ofPrimitive(Primitive.ofBooleanValue(boolean))
          ),
          LiteralType.ofSimpleType(SimpleType.BOOLEAN)
        )
      case long: Long =>
        (
          BindingData.ofScalar(
            Scalar.ofPrimitive(Primitive.ofIntegerValue(long))
          ),
          LiteralType.ofSimpleType(SimpleType.INTEGER)
        )
      case double: Double =>
        (
          BindingData.ofScalar(
            Scalar.ofPrimitive(Primitive.ofFloatValue(double))
          ),
          LiteralType.ofSimpleType(SimpleType.FLOAT)
        )
      case instant: Instant =>
        (
          BindingData.ofScalar(
            Scalar.ofPrimitive(Primitive.ofDatetime(instant))
          ),
          LiteralType.ofSimpleType(SimpleType.DATETIME)
        )
      case duration: Duration =>
        (
          BindingData.ofScalar(
            Scalar.ofPrimitive(Primitive.ofDuration(duration))
          ),
          LiteralType.ofSimpleType(SimpleType.DURATION)
        )
      case list: Seq[_] =>
        val literalType = literalTypeOpt.getOrElse {
          val (_, innerLiteralType) = toBindingData(
            list.headOption.getOrElse(
              throw new RuntimeException(
                "Can't create binding for an empty list without knowing the type, use SdkBindingData.of<type>Collection(...)"
              )
            ),
            literalTypeOpt = None
          )

          LiteralType.ofCollectionType(innerLiteralType)
        }

        (
          BindingData.ofCollection(
            list
              .map { innerValue =>
                val (bindingData, _) = toBindingData(innerValue, literalTypeOpt)
                bindingData
              }
              .toList
              .asJava
          ),
          literalType
        )
      case map: Map[String, _] =>
        val literalType = literalTypeOpt.getOrElse {
          val (_, innerLiteralType) = toBindingData(
            map.headOption
              .map(_._2)
              .getOrElse(
                throw new RuntimeException(
                  "Can't create binding for an empty map without knowing the type, use SdkBindingData.of<type>Map(...)"
                )
              ),
            literalTypeOpt = None
          )

          LiteralType.ofMapValueType(innerLiteralType)
        }
        (
          BindingData.ofMap(
            map
              .mapValues { innerValue =>
                val (bindingData, _) = toBindingData(innerValue, literalTypeOpt)
                bindingData
              }
              .toMap
              .asJava
          ),
          literalType
        )
      case other =>
        throw new IllegalStateException(
          s"${other.getClass.getSimpleName} class is not supported as SdkBindingData inner class"
        )
    }
  }

  private def createSdkBindingData[T](
      value: T,
      literalTypeOpt: Option[LiteralType] = None
  ): SdkBindingData[T] = {
    val (bindingData, literalType) = toBindingData(value, literalTypeOpt)
    SdkBindingData.create(bindingData, literalType, value)
  }
}
